### DISCUSSION TITLE
Problems with inlineSyncPlugin when customizing nodes

### DISCUSSION THREAD

**Discussion by iHaPBoy** [Milkdown/milkdown#835] [General]
I'm using JSON format to store data for the editor. I need to implement the mention feature and need to add a custom `prosemirror node` (with some attrs), but don't want to add the `remark node type` (don't want to extend the markdown syntax, markdown is only used as import, export).

I found that I had to implement the `parseMarkdown`, and more importantly, the `inlineSyncPlugin` would convert in real time as the content was entered, causing the custom `prosemirror node` to be replaced.

I think `inlineSyncPlugin` limits the extensibility, would it be better to use the `Input Rule` approach before? [Inline sync plugin #695](https://github.com/Saul-Mirone/milkdown/pull/695)

Milkdown has great support for Markdown, including syntax, input, and output. I want to store it in JSON format and extend it with more node types (not necessarily with markdown node types). Any ideas?

@Saul-Mirone

**Comment by Saul-Mirone** [comment]
I think you have to add `toMarkdown` and `parseMarkdown` for this case.

The problem of `InputRule` is that it cannot detect the content after the cursor, which will cause the prosemirror state and remark AST not equal.

The good news for you is that you can disable it by config the `shouldSyncNode` here to always return `false`.
https://github.com/Saul-Mirone/milkdown/blob/main/packages/preset-commonmark/src/plugin/inline-sync/config.ts#L37

But I still recommend you to implement the parseMarkdown and toMarkdown part, it won't be complicated.

**Reply by iHaPBoy** [reply]
> I think you have to add `toMarkdown` and `parseMarkdown` for this case.

The prosemirror node `mention` contains some attrs, such as user id, username. Does this mean I have to implement a remark node type or markdown syntax to include these attributes?

**Reply by Saul-Mirone** [reply]
Yes, it should have some kind of reflection between markdown and prosemirror state.
